File: src/web/magmac/api/iter/collect/ResultCollector.ts
import { Tuple2 } from "../../../../magmac/api/Tuple2";
import { Ok } from "../../../../magmac/api/result/Ok";
import { Result } from "../../../../magmac/api/result/Result";
export class ResultCollector<T,  C,  X> {
	createInitial() : Result<C, X> {return new Ok<>( this.collector.createInitial( ));;}
	fold(currentResult : Result<C, X>, element : Result<T, X>) : Result<C, X> {return currentResult.and( 0).mapValue( 0);;}
}

File: packages/compiler-java/src/main/java/magmac/api/iter/collect/ResultCollector.java
package magmac.api.iter.collect;

import magmac.api.Tuple2;
import magmac.api.result.Ok;
import magmac.api.result.Result;

/**
 * Collector that transforms {@link magmac.api.result.Result} values while accumulating.
 */

public record ResultCollector<T, C, X>(
        Collector<T, C> collector
) implements Collector<Result<T, X>, Result<C, X>> {
    @Override
    public Result<C, X> createInitial() {
        return new Ok<>(this.collector.createInitial());
    }

    @Override
    public Result<C, X> fold(Result<C, X> currentResult, Result<T, X> element) {
        return currentResult.and(() -> element)
                .mapValue((Tuple2<C, T> tuple) -> this.collector.fold(tuple.left(), tuple.right()));
    }
}
